src/api/v1/reports.ts

```diff
@@ -0,0 +1,107 @@
+import { Hono } from "hono";
+import * as vocab from "@fedify/fedify/vocab";
+
+import { z } from "zod";
+import { db } from "../../db";
+import { zValidator } from "@hono/zod-validator";
+
+import federation from "../../federation";
+import { type Variables, scopeRequired, tokenRequired } from "../../oauth";
+import { accounts, posts, reports, type Post, type Report } from "../../schema";
+import { uuidv7 } from "uuidv7-js";
+import { serializeReport } from "../../entities/report";
+import { eq, inArray } from "drizzle-orm";
+
+const app = new Hono<{ Variables: Variables }>();
+
+const reportSchema = z.object({
+  comment: z.string().trim().min(1).max(1000).optional(),
+  account_id: z.string().uuid(),
+  status_ids: z.array(z.string().uuid()).min(1).optional(),
+  // discarded by defined by the Mastodon API:
+  category: z.string().optional(),
+  rule_ids: z.string().array().optional(),
+  forward: z.boolean().optional()
+});
+
+app.post(
+  "/",
+  tokenRequired,
+  scopeRequired(["write:reports"]),
+  zValidator(
+    "json",
+    reportSchema
+  ),
+  async (c) => {
+    const accountOwner = c.get("token").accountOwner;
+    if (accountOwner == null) {
+      return c.json(
+        { error: "This method requires an authenticated user" },
+        422,
+      );
+    }
+
+    const data = c.req.valid("json");
+
+    const targetAccount = await db.query.accounts.findFirst({
+      where: eq(accounts.id, data.account_id),
+      with: { owner: true, successor: true }
+    });
+
+    if (targetAccount == null) {
+      return c.json({ error: "Record not found" }, 404);
+    }
+
+    let targetPosts: Post[] = [];
+    if (data.status_ids != null && data.status_ids.length > 0) {
+      targetPosts = await db.query.posts.findMany({
+        where: inArray(posts.id, data.status_ids)
+      })
```

Comment:
There doesn't seem to be anything in the database to indicate whether or not an Post is local to this server or a remote server? If there was, then we could assert that all the `targetPosts` are indeed remote.

We could also limit the where to include the `actor_id` equal to the `targetAccount.id` to ensure we only report posts from the author who created them (i.e., you can't report Jane about Alice's post)

Reply:
> There doesn't seem to be anything in the database to indicate whether or not an Post is local to this server or a remote server? If there was, then we could assert that all the `targetPosts` are indeed remote.

Yeah, as you said, there's nothing to indicate whether a post is remote or local in the `posts` table itself—instead, you need to outer-join it to the `account_owners` table… Well, it's kinda complicated.

> We could also limit the where to include the `actor_id` equal to the `targetAccount.id` to ensure we only report posts from the author who created them (i.e., you can't report Jane about Alice's post)

I think we can simply add a filter to the query?

```suggestion
      targetPosts = await db.query.posts.findMany({
        where: and(
          inArray(posts.id, data.status_ids),
          eq(posts.account_id, targetAccount.id)
        )
      })
```

Reply:
Yeah, the local vs remote posts thing isn't such an issue if we constrain to posts by the `targetAccount.id`.